### PR TITLE
feature/560 expiry selector

### PR DESCRIPTION
**Pull request by coltraneyan**
https://user-images.githubusercontent.com/80134686/179375430-ed7be104-8b5a-4b65-986a-8f140dc74c9f.mov
- takes in updateExpiry function to pass state back to parent (output in milliseconds, e.g. 840000, date can be handed at sign so expiry time is exact)
- 1 is default text input, NaN input will revert to 1 in output function. (otherwise, this error can be handled at sign time)

### CODE REVIEW

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.styles.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import styled, { css } from "styled-components/macro";
+
+import { InputOrButtonBorderStyleType2 } from "../../../../style/mixins";
+
+export const SelectorStyle = css`
+  ${InputOrButtonBorderStyleType2}
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.styles.tsx:6]:
```suggestion
  ${InputOrButtonBorderStyleType2};
```

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.styles.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+`;
+
+export const Wrapper = styled.div`
+  display: flex;
+  height: 2rem;
+  width: fit-content;
+`;
+
+export const Title = styled.button`
+  ${SelectorStyle}
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.styles.tsx:20]:
```suggestion
  ${SelectorStyle};
```

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.styles.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  :focus {
+    border: 1px solid ${({ theme }) => theme.colors.borderGrey};
+    cursor: initial;
+  }
+`;
+
+export const Input = styled.input`
+  all: unset;
+
+  ${SelectorStyle}
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.styles.tsx:38]:
```suggestion
  ${SelectorStyle};
```

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  {
+    label: "weeks",
+    value: `${7 * 24 * 60 * MS_PER_MINUTE}`,
+  },
+];
+
+const floatRegExp = new RegExp("^([0-9])*$");
+
+export type ExpirySelectorProps = {
+  updateExpiry: (msToExpiry: number) => void;
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:30]:
```suggestion
    onChange: (expiryDate: number) => void;
```
Value returned should be current date + expiry time

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+};
+
+export const ExpirySelector: React.FC<ExpirySelectorProps> = ({
+  updateExpiry,
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:34]:
```suggestion
  onChange,
```

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+}) => {
+  const [unit, setUnit] = useState(options[0]);
+  const [amount, setAmount] = useState("1");
+
+  useEffect(() => {
+    updateExpiry((parseInt(amount) || 1) * parseInt(unit.value));
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:40]:
```suggestion
    const msToExpiry = (parseInt(amount, 0) || 0) * parseInt(unit.value, 0);
    onChange(new Date().getTime() + msToExpiry);

```

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  }, [unit, amount]);
+
+  function handleUnitChange(option: SelectOption) {
+    setUnit(option);
+  }
+
+  function handleAmountChange(value: string) {
+    if (value === "" || floatRegExp.test(value)) {
+      value = value.replace(/^0+/, "0");
+      setAmount(value);
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:50]:
```suggestion
  function handleAmountChange(e: React.ChangeEvent<HTMLInputElement>) {
    const { value } = e.target;
    if (value === "" || floatRegExp.test(value)) {
      setAmount(value.replace(/^0+/, "0"));
```

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  return (
+    <Wrapper>
+      <Title>EXPIRES IN</Title>
+      <Input
+        maxLength={3}
+        value={amount}
+        onChange={(e) => handleAmountChange(e.target.value)}
+      ></Input>
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:61]:
```suggestion
        onChange={handleAmountChange}
      />
```

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      <Dropdown
+        value={unit}
+        options={options}
+        onChange={handleUnitChange}
+      ></Dropdown>
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:66]:
```suggestion
      />
```

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  function handleAmountChange(value: string) {
+    if (value === "" || floatRegExp.test(value)) {
+      value = value.replace(/^0+/, "0");
+      setAmount(value);
+    }
+  }
+
+  return (
+    <Wrapper>
+      <Title>EXPIRES IN</Title>
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:56]: Add translation for this. You can add it temporarily in locales/en/translation.json. Did you get setup yet at poeditor? That's where we manage all translations.

**Review by piersss — CHANGES_REQUESTED** [review]
Noice.

Looks good! Also make sure to check out light mode, that's easy to overlook.

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx (hunk 10 above)
Comment (coltraneyan) [reply src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:56]: not set up in poeditor. ill ping don.

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+export const ExpirySelector: React.FC<ExpirySelectorProps> = ({ onChange }) => {
+  const { t } = useTranslation();
+  const [unit, setUnit] = useState(options[0]);
+  const [amount, setAmount] = useState("1");
+
+  useEffect(() => {
+    const msToExpiry = (parseInt(amount, 0) || 0) * parseInt(unit.value, 0);
+    onChange(new Date().getTime() + msToExpiry);
+  }, [unit, amount]);
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:42]:
```suggestion
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [unit, amount]);
```

Comment (piersss) [reply src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:42]: Got a warning from this, didn't notice last time.

**Review by piersss — CHANGES_REQUESTED** [review]
One small thing

File: src/components/MakeWidget/subcomponents/ExpirySelector/helpers/getExpiryUnitText.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    case "hours":
+      return t("common.hours");
+    case "days":
+      return t("common.days");
+    case "weeks":
+      return t("common.weeks");
+    default:
+      return t("common.unknown");
+  }
+}
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/helpers/getExpiryUnitText.ts:16]:
```
const MS_PER_MINUTE = 60000;

export default function getExpirySelectOptions(unit: String, t: TFunction): SelectOption[] {
  return [
    {
      label: t("common.minutes"),
      value: `${MS_PER_MINUTE}`,
    },
    {
      label: t("common.hours"),
      value: `${60 * MS_PER_MINUTE}`,
    },
    {
      label: t("common.minutes"),
      value: `${24 * 60 * MS_PER_MINUTE}`,
    },
    {
      label: t("common.weeks"),
      value: `${7 * 24 * 60 * MS_PER_MINUTE}`,
    },
  ];
}
```

Comment (piersss) [reply src/components/MakeWidget/subcomponents/ExpirySelector/helpers/getExpiryUnitText.ts:16]: This makes more sense!

File: src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx [hunk 13]
```diff
@@ -33,7 +34,15 @@ export type ExpirySelectorProps = {
...
 export const ExpirySelector: React.FC<ExpirySelectorProps> = ({ onChange }) => {
   const { t } = useTranslation();
-  const [unit, setUnit] = useState(options[0]);
+
+  const translatedOptions = useMemo(() => {
+    return options.map((option) => ({
+      label: getExpiryUnitText(option.label, t),
+      value: option.value,
+    }));
+  }, [options, t]);
```
Comment (piersss) [inline src/components/MakeWidget/subcomponents/ExpirySelector/ExpirySelector.tsx:43]: Dependency warning again.

**Review by piersss — CHANGES_REQUESTED** [review]

**Pull request merged**